sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -628,24 +623,59 @@ case class Substring(str: Expression, pos: Expression, len: Expression)
 
   override def eval(input: InternalRow): Any = {
     val string = str.eval(input)
-    val po = pos.eval(input)
-    val ln = len.eval(input)
-
-    if ((string == null) || (po == null) || (ln == null)) {
-      null
-    } else {
-      val start = po.asInstanceOf[Int]
-      val length = ln.asInstanceOf[Int]
-      string match {
-        case ba: Array[Byte] =>
-          val (st, end) = slicePos(start, length, () => ba.length)
-          ba.slice(st, end)
-        case s: UTF8String =>
+    if (string != null) {
+      val po = pos.eval(input)
+      if (po != null) {
+        val ln = len.eval(input)
+        if (ln != null) {
+          val start = po.asInstanceOf[Int]
+          val length = ln.asInstanceOf[Int]
+          val s = string.asInstanceOf[UTF8String]
```

Comment:
slicepos is going to be really bad for performance, since it uses pattern matching and creates tuples. can we remove the function?